core/src/main/java/jenkins/console/ConsoleUrlProvider.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2023, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package jenkins.console;
+
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.Functions;
+import hudson.model.Describable;
+import hudson.model.Descriptor;
+import hudson.model.Run;
+import hudson.model.User;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import jenkins.model.Jenkins;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.Stapler;
+
+/**
+ * Extension point that allows implementations to redirect build console links to a specified URL.
+ * <p>In order to produce links to console URLs in Jelly templates, use {@link Functions#getConsoleUrl}.
+ * @see Functions#getConsoleUrl
+ * @since TODO
+ */
+public interface ConsoleUrlProvider extends Describable<ConsoleUrlProvider> {
+    @Restricted(NoExternalUse.class)
+    Logger LOGGER = Logger.getLogger(ConsoleUrlProvider.class.getName());
```

Comment:
Since `ConsoleUrlProvider` is an interface, the field can't be `private` and it is `public static final` by default. I think we could use a dynamic `private Logger getLogger()` method that calls `Logger.getLogger` each time to avoid the field though if preferred (not sure how costly it is to call `Logger.getLogger` on demand).